pyFAI/ext/CSC_common.pxi

```diff
@@ -0,0 +1,308 @@
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+"""Common CSR integrator"""
+
+__author__ = "Jérôme Kieffer"
+__contact__ = "Jerome.kieffer@esrf.fr"
+__date__ = "08/09/2022"
+__status__ = "stable"
+__license__ = "MIT"
+
+import cython
+import numpy
+
+from ..containers import Integrate1dtpl, Integrate2dtpl, ErrorModel
+
+
+cdef class CscIntegrator(object):
+    """Abstract class which implements only the integrator...
+
+    Now uses CSR (Compressed Sparse Column) with main attributes:
+    * nnz: number of non zero elements
+    * data: coefficient of the matrix in a 1D vector of float32
+    * indices: Column index position for the data (same size as
+    * indptr: Column pointer indicates the start of a given Column. len ncol+1
+
+    Nota: nnz = indptr[-1]+1 = len(indices) = len(data)
+    """
+    cdef:
+        readonly index_t input_size, output_size, nnz
+        readonly data_t empty
+        readonly data_t[::1] _data
+        readonly index_t[::1] _indices, _indptr
+
+    def __init__(self,
+                  tuple lut,
+                  int input_size,
+                  int output_size,
+                  data_t empty=0.0):
+
+        """Constructor for a CSR generic integrator
+
+        :param lut: Sparse matrix in CSR format, tuple of 3 arrays with (data, indices, indptr)
+        :param size: input image size
+        :param empty: value for empty pixels
+        """
+        self.empty = empty
+        assert input_size >= len(lut[2])-1, "image size is OK"
+        assert len(lut) == 3, "Sparse matrix is expected as 3-tuple CSR with (data, indices, indptr)"
+        assert len(lut[1]) == len(lut[0]),  "Sparse matrix in CSC format is expected to have len(data) == len(indices) is expected as 3-tuple CSR with (data, indices, indptr)"
+        self._data = numpy.ascontiguousarray(lut[0], dtype=data_d)
+        self._indices = numpy.ascontiguousarray(lut[1], dtype=numpy.int32)
+        self._indptr = numpy.ascontiguousarray(lut[2], dtype=numpy.int32)
+        self.nnz = len(lut[1])
+        self.output_size = output_size
+        self.input_size = input_size
+
+    def __dealloc__(self):
+        self._data = None
+        self._indices = None
+        self._indptr = None
+        self.empty = 0
+        self.input_size = 0
+        self.output_size = 0 
+        self.nnz = 0
+
+    @property
+    def data(self):
+        return numpy.asarray(self._data)
+    @property
+    def indices(self):
+        return numpy.asarray(self._indices)
+    @property
+    def indptr(self):
+        return numpy.asarray(self._indptr)
+
+
+    def integrate_ng(self,
+                     weights,
```

Comment:
Are these the raw pixel intensities, before the dark/flat/mask etc. For 'weights' I would think about something else.

Reply:
Yes, those are the raw pixel intensities. The name "weights" comes from the time when intensities were used as weights for histogramming the pixel position.